### PR TITLE
codeintel: use COUNT OVER() window function for getting uploads/indexes total count

### LINKED ISSUE #34915: codeintel: Improve queries that require a distinct `COUNT(*)`
In [several places in the codeintel dbstore/lsifstore](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+f:codeintel+f:store+countQuery&patternType=literal), we do the following:

- start a transaction
- run a base query to count total rows (with filter but without limit/offset)
- run a query returning wanted data (with filter and limit/offset)

We can possibly run these queries together (see [this SO question](https://stackoverflow.com/questions/28888375/run-a-query-with-a-limit-offset-and-also-get-the-total-number-of-rows)) with a window function. This may reduce the total cost of two queries with one slightly more expensive query.

Each of these cases may see a performance boost. The traces I've seen when optimizing #32328 leads me to believe that there is some non-trivial overhead in these separate count queries (200ms for a count + 200ms for the real query vs possibly 300ms for a single query).

### PR DESCRIPTION
**Pull request by Strum355**
May or may not close #34915 by implementing the approach from the linked stackoverflow issue. Performance improvements weren't stable (sometimes faster, sometimes on-par), so I want to observe this in production for a while to see how the average case does, given we're only having to make one query here instead of two, the overhead saved may work in its favour.

## Test plan

Covered by existing tests and dashboards

### CODE REVIEW

File: internal/database/basestore/rows.go [hunk 1]
```diff
@@ -53,6 +53,31 @@ func NewSliceScanner[T any](f func(dbutil.Scanner) (T, error)) func(rows *sql.Ro
 	}
 }
 
+// NewSliceWithCountScanner returns a basestore scanner function that returns all
+// the values of the query result, as well as the count from the `COUNT(*) OVER()`
+// window function. The given function is invoked multiple times with a SQL rows
+// object to scan a single value.
```
Comment (efritz) [inline internal/database/basestore/rows.go:59]: I have more context on this particular query than most readers will; could we add a representative query + use that this is meant to work with? A trivial example with some canned fields would be great.

File: enterprise/internal/codeintel/stores/dbstore/uploads.go [hunk 2]
```diff
@@ -88,9 +88,48 @@ func scanUpload(s dbutil.Scanner) (upload Upload, _ error) {
...
+		&upload.NumFailures,
+		&upload.RepositoryID,
+		&upload.RepositoryName,
+		&upload.Indexer,
+		&dbutil.NullString{S: &upload.IndexerVersion},
+		&upload.NumParts,
+		pq.Array(&rawUploadedParts),
+		&upload.UploadSize,
+		&upload.AssociatedIndexID,
+		&upload.Rank,
```
Comment (efritz) [inline enterprise/internal/codeintel/stores/dbstore/uploads.go:114]: Wonder if there's an easy way to share the slice of pointers here with the regular scan function 🤔

**Review by efritz — APPROVED** [review]
If unit tests pass, then :shipit:

File: internal/database/basestore/rows.go (hunk 1 above)
Comment (Strum355) [reply internal/database/basestore/rows.go:59]: What do you think of the updated one?

File: enterprise/internal/codeintel/stores/dbstore/uploads.go (hunk 2 above)
Comment (Strum355) [reply enterprise/internal/codeintel/stores/dbstore/uploads.go:114]: Dont think so unfortunately, not without going reflect

**Pull request merged**